lib/src/kbasesearchengine/common/GUID.java

```diff
@@ -128,21 +152,54 @@ public GUID(String textGUID) {
             this.subObjectId = subObjectPart.substring(slashPos + 1);
         }
     }
-    
-    //TODO NOW if the version or accessGroupID is missing, how do you know which is which? Need to know storage code -> has access group mapping or have a placeholder (-1) for missing access group
+
+    /**
+     * Returns the URL-encoded version of this GUID.
+     *
+     * For the sake of readability the URL-encoded format of the GUID is,
+     *
+     *   {storageCode}___[{AccessGroupID}_]{accessGroupObjectId}[_{version}][___{subObjectType}_{subObjectId}],
+     *
+     *  where special characters in the individual tokens will get URL-encoded.
+     *
+     * @return URL-encoded version of this GUID
+     */
+    public String getURLEncoded() throws IOException {
+
+        String encodedGUID = toString("___", "_");
+
+        try {
+
+             encodedGUID =  URLEncoder.encode(encodedGUID, "UTF-8");
```

Comment:
I'm not clear on why we need to worry about urlencoding on the server side - that's a client side issue, right?

Reply:
KBSearch is a client to ElasticSearch. We use the ES REST API to interact with ES.

Reply:
yeah, that didn't make sense. What I'm trying to say is that I don't think the stored ID needs to be urlencoded - the client (in this case the ES REST client) will handle the urlencoding when it sends the request, and the ES server will unencode it before presenting it to the business logic. That's the way I'd expect it to work anyway.